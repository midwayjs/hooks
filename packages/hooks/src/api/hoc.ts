import {
  Api,
  ApiFunction,
  HooksMiddleware,
  Middleware,
  validateArray,
  validateFunction,
} from '@midwayjs/hooks-core'

type Controller = {
  middleware?: HooksMiddleware[]
}

const { deprecate } = require('util')

/**
 * @deprecated Use `Api(Middleware(...middlewares))` instead
 */
export const withController = deprecate(
  <T extends ApiFunction>(controller: Controller, func: T) => {
    if (controller?.middleware !== undefined) {
      validateArray(controller.middleware, 'controller.middleware')
    }
    validateFunction(func, 'func')
    return Api(Middleware(controller.middleware), func)
  },
  'withController is deprecated. Use `Api(Middleware(...middlewares))` instead.'
)

/**
 * @deprecated Use `Api(Middleware(...middlewares))` instead
 */
export const withMiddleware = deprecate(
  <T extends ApiFunction>(middleware: HooksMiddleware[], func: T) => {
    validateArray(middleware, 'middleware')
    validateFunction(func, 'func')
    return Api(Middleware(middleware), func)
  },
  'withMiddleware is deprecated. Use `Api(Middleware(...middlewares))` instead.'
)
