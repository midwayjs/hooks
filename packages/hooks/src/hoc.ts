import {
  ApiFunction,
  HooksMiddleware,
  validateArray,
  validateFunction,
} from '@midwayjs/hooks-core'

type Controller = {
  middleware?: HooksMiddleware
}

export function withController<T extends ApiFunction>(
  controller: Controller,
  func: T
) {
  if (controller?.middleware !== undefined) {
    validateArray(controller.middleware, 'controller.middleware')
  }
  validateFunction(func, 'func')

  const withControllerProxy: ApiFunction = async function withControllerProxy(
    ...args: any[]
  ) {
    return func.apply(this, args)
  }

  withControllerProxy.middleware = controller.middleware

  return withControllerProxy as T
}

export function withMiddleware<T extends ApiFunction>(
  middleware: HooksMiddleware,
  func: T
) {
  validateArray(middleware, 'middleware')
  validateFunction(func, 'func')

  const proxy: ApiFunction = async (...args: any[]) => func.apply(this, args)
  proxy.middleware = middleware
  return proxy as T
}
