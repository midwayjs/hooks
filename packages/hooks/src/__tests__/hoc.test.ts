import { withController, withMiddleware } from '../hoc'

const noop = async () => {}

test('expect wrapped fn to be called', () => {
  const fn = jest.fn()
  const target = withController(
    {
      middleware: [],
    },
    fn
  )
  target()
  expect(fn).toHaveBeenCalledTimes(1)

  const target2 = withMiddleware([], fn)
  target2()
  expect(fn).toHaveBeenCalledTimes(2)
})

test('validate arguments', () => {
  expect(withController).toThrowErrorMatchingSnapshot()
  expect(withMiddleware).toThrowErrorMatchingSnapshot()

  expect(() =>
    withController({ middleware: null }, noop)
  ).toThrowErrorMatchingSnapshot()
  expect(() => withController({}, null)).toThrowErrorMatchingSnapshot()

  expect(() => withMiddleware(null, null)).toThrowErrorMatchingSnapshot()
  expect(() => withMiddleware([], null)).toThrowErrorMatchingSnapshot()
})
