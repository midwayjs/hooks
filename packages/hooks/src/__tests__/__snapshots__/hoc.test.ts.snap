// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validate arguments 1`] = `"[ERR_INVALID_ARG_TYPE]: The 'func' argument must be of type Function. Received undefined"`;

exports[`validate arguments 2`] = `"[ERR_INVALID_ARG_TYPE]: The 'middleware' argument must be of type Array. Received undefined"`;

exports[`validate arguments 3`] = `"[ERR_INVALID_ARG_TYPE]: The 'controller.middleware' argument must be of type Array. Received null"`;

exports[`validate arguments 4`] = `"[ERR_INVALID_ARG_TYPE]: The 'func' argument must be of type Function. Received null"`;

exports[`validate arguments 5`] = `"[ERR_INVALID_ARG_TYPE]: The 'middleware' argument must be of type Array. Received null"`;

exports[`validate arguments 6`] = `"[ERR_INVALID_ARG_TYPE]: The 'func' argument must be of type Function. Received null"`;
